osu.Game/Online/Chat/MessageNotifier.cs

```diff
@@ -0,0 +1,234 @@
+        private Bindable<User> localUser;
+
+        /// <summary>
+        /// Determines if the user is able to see incoming messages.
+        /// </summary>
+        public bool IsActive => chatOverlay?.IsPresent == true;
+
+        private readonly List<PrivateMessageNotification> privateMessageNotifications = new List<PrivateMessageNotification>();
+
+        [BackgroundDependencyLoader]
+        private void load(OsuConfigManager config, IAPIProvider api)
+        {
+            notifyOnMention = config.GetBindable<bool>(OsuSetting.ChatHighlightName);
+            notifyOnChat = config.GetBindable<bool>(OsuSetting.ChatMessageNotification);
+            localUser = api.LocalUser;
+
+            // Listen for new messages
+            channelManager.JoinedChannels.ItemsAdded += joinedChannels =>
+            {
+                foreach (var channel in joinedChannels)
+                    channel.NewMessagesArrived += channel_NewMessagesArrived;
+            };
+
+            channelManager.JoinedChannels.ItemsRemoved += leftChannels =>
+            {
+                foreach (var channel in leftChannels)
+                    channel.NewMessagesArrived -= channel_NewMessagesArrived;
+            };
+        }
+
+        private void channel_NewMessagesArrived(IEnumerable<Message> messages)
+        {
+            if (messages == null || !messages.Any())
+                return;
+
+            HandleMessages(messages.First().ChannelId, messages);
+        }
+
+        /// <remarks>
+        /// Resolves the channel id
+        /// </remarks>
+        public void HandleMessages(long channelId, IEnumerable<Message> messages)
+        {
+            var channel = channelManager.JoinedChannels.FirstOrDefault(c => c.Id == channelId);
+
+            if (channel == null)
+            {
+                Logger.Log($"Couldn't resolve channel id {channelId}", LoggingTarget.Information);
+                return;
+            }
+
+            HandleMessages(channel, messages);
+        }
+
+        public void HandleMessages(Channel channel, IEnumerable<Message> messages)
+        {
+            // don't show if the ChatOverlay and the target channel is visible.
+            if (IsActive && channelManager.CurrentChannel.Value == channel)
+                return;
+
+            foreach (var message in messages)
+            {
+                // ignore messages that already have been read
+                if (message.Id < channel.LastReadId)
+                    return;
+
+                // ignore messages from yourself
+                var localUsername = localUser.Value.Username;
+
+                if (message.Sender.Username == localUsername)
+                    continue;
+
+                if (checkForPMs(channel, message))
+                    continue;
+
+                // change output to bool again if another "message processor" is added.
+                checkForMentions(channel, message, localUsername);
```

Comment:
This seems plain wrong. Suppose you're in a normal channel (non-PM). `checkForPMs` will always return `false`, therefore making mentions not work at all?

Reply:
> but mentions still won't work in PM channels, which they probably should do.

It's like mentioning your friend in an one to one DM in Discord. It doesn't make much sense (in my opinion), sending two notifications about the same message with the same destination (your friends PMs).

Reply:
Don't necessarily disagree with your conclusion either. The most painful thing about this to me is that the order of checking now silently matters and will generate regressions / unexpected behaviour if another conditional is improperly inserted. The processor proposal would help isolate concerns here.

Reply:
Is it fine if I include a warning comment, for the lifetime of this pull request living inside lazer?

Reply:
A comment wouldn't hurt if this is to stay as is. You could also describe the rationale behind this particular ordering (why PMs first, etc.)